### ISSUE TITLE
x64 Checked Windows_NT GCStress=0xc has four failures

### ISSUE THREAD

**Issue by tannergooding** [dotnet/coreclr#21145]
#21128 fixed an issue where the GCStress tests were not running with the appropriate timeout. However, there were still four remaining failures (see https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_windows_nt_gcstress0xc_prtest/85/ for more details):
* Interop_COM._NativeClients_Primitives_Primitives_._NativeClients_Primitives_Primitives_cmd
* Loader_classloader._TypeGeneratorTests_TypeGeneratorTest1178_Generated1178_Generated1178_._TypeGeneratorTests_TypeGeneratorTest1178_Generated1178_Generated1178_cmd
* Loader_classloader._TypeGeneratorTests_TypeGeneratorTest856_Generated856_Generated856_._TypeGeneratorTests_TypeGeneratorTest856_Generated856_Generated856_cmd
* Loader_classloader._TypeGeneratorTests_TypeGeneratorTest1070_Generated1070_Generated1070_._TypeGeneratorTests_TypeGeneratorTest1070_Generated1070_Generated1070_cmd

The first was a timeout and the latter three looked like actual failures. All of these should be investigated further.

**Comment by sandreenko** [comment]
I was able to catch this failure in a debugger and get the stack trace:
```
Assert failure(PID 41436 [0x0000a1dc], Thread: 45556 [0xb1f4]): !CREATE_CHECK_STRING(pMT && pMT->Validate())

CORECLR! Object::ValidateInner + 0x3EB (0x00007ffb`b3cc14fb)
CORECLR! Object::Validate + 0x13A (0x00007ffb`b3cc0fea)
CORECLR! GcInfoDecoder::ReportRegisterToGC + 0x123 (0x00007ffb`b44242df)
CORECLR! GcInfoDecoder::ReportSlotToGC + 0xCD (0x00007ffb`b442451d)
CORECLR! GcInfoDecoder::EnumerateLiveSlots + 0xCBE (0x00007ffb`b4423732)
CORECLR! EECodeManager::EnumGcRefs + 0x3A3 (0x00007ffb`b4088803)
CORECLR! GcStackCrawlCallBack + 0x2F6 (0x00007ffb`b4333e16)
CORECLR! Thread::MakeStackwalkerCallback + 0x52 (0x00007ffb`b3df8d92)
CORECLR! Thread::StackWalkFramesEx + 0x180 (0x00007ffb`b3dfa91c)
CORECLR! Thread::StackWalkFrames + 0x184 (0x00007ffb`b3dfa6fc)
    File: f:\git\coreclr\src\vm\object.cpp Line: 702
    Image: F:\git\coreclr\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\CoreRun.exe
```

**Comment by sandreenko** [comment]
Has anybody seen this assert/stack before @jkotas, @Maoni0, @AndyAyersMS ?

I was running `Windows_NT.x64.Checked\baseservices\threading\paramthreadstart\ThreadStartNeg4` in 20 threads on 4 cores (8 virtual) machine to get the repro.

**Comment by AndyAyersMS** [comment]
If you search you can find examples in various issues, eg #17387, #19127

Probably a bug in GC info produced by the jit; you should figure out which managed stack frame is being examined by the GC stackwalk -- that is likely where the bug happens.

**Comment by sandreenko** [comment]
`CORECLR! Object::ValidateInner` shows that `this` is `0x000001f2`000a2aa8 class Object *` and GCRoot for it shows: 
```
    000000a67dafd0b0 00007ffbb8431dac System.Diagnostics.Tracing.EventSource.Initialize(System.Guid, System.String, System.String[])
        rsi: 
            ->  000001f2000a2aa8 <unknown>

```
then if you look at `00007ffbb8431dac ` you will find the offset where the issue happens:
```
Method Name:          System.Diagnostics.Tracing.EventSource.Initialize(System.Guid, System.String, System.String[])
Class:                00007ffbb77af248
MethodTable:          00007ffbb8803af8
mdToken:              0000000006004ba5
Module:               00007ffbb7671000
IsJitted:             yes
Current CodeAddr:     00007ffbb8431c40
Code Version History:
  CodeAddr:           00007ffbb8431c40  (Non-Tiered)
  NativeCodeVersion:  0000000000000000

```
and the offset is `0x16c`.

```
IN004a: 000160 mov      rcx, rdi
IN004b: 000163 mov      rdx, gword ptr [V00 rbp+10H]
IN004c: 000167 call     System.Diagnostics.Tracing.EventProvider:Register(ref):this
IN004d: 00016C mov      rcx, gword ptr [V00 rbp+10H]
```

I have checked jit dump for `System.Diagnostics.Tracing.EventSource.Initialize` and `rsi` is alive at this offset, but windbg shows that memory `000001f2000a2aa8 ` is empty.

Does not it look like a heap location that was moved but GC did not update the register value with its address?

**Comment by AndyAyersMS** [comment]
Can you dump the gc info for the method when debugging (via sos) and see if it matches the gc info the jit thinks it produced?

**Comment by tommcdon** [comment]
@jorive this looks related to #21461

**Comment by sandreenko** [comment]
Ok, looks like it is time to add some description here.
To hit this issue you need several conditions:
1. `EventPipe` tracing thread that wakes up every `DisabledPollingIntervalMilliseconds`;
2. CPU under a heavy load that finishs to run a small app and calls `UnloadAppDomain` and it calls `EEShutDown`;
3. When the main thread (with shutdown) is running `EventPipe` should start running a new thread that runs `System.Diagnostic.Tracing.EventSource.Initialize`; **<- issue N1.**

That situation is already wrong, because we should not run any managed code after we called shutdown.
As I understand it will be fixed with #21461.

But to get AV as we get in this issue we need more:
4. `System.Diagnostic.Tracing.EventSource.Initialize` should call `eventPipeProvider.Register(this);` that throws `throw new ArgumentException(Convert.ToString(unchecked((int)status)));` and goes into a funclet;
4. we need `GCStress=0xc` so `new` will hit `hlt` instruction (that is GC stress marker) and handle it with `vm\exceptionhandling.cpp::ProcessCLRException` that calls `IsGcMarker` and sometimes it can return `false` because part of it was already killed by shutdown; **<- issue N2**
5. if it happens (`IsGcMarker` returned false when it should return `true`) then we think that we hit a real SEH exception and we start to process with another funclet;
6. `CreateThrowable` should call a regular GC and this GC should find a root that was moved away and not updated, that was `rsi` when we entered `eventPipeProvider.Register(this)` and it pointed to `var eventPipeProvider = new OverideEventProvider(this, EventProviderType.EventPipe);` and because `GcInfoDecoder::ReportRegisterToGC `  is called with an incorrect pointer without method table and we hit `Assert failure(PID 41436 [0x0000a1dc], Thread: 45556 [0xb1f4]): !CREATE_CHECK_STRING(pMT && pMT->Validate())` **<- issue N3.**

**Comment by sandreenko** [comment]
I am working on reproing **issues 2 and 3** without **issue 1**, but do not get any luck with it.

**Comment by jkotas** [comment]
> I understand it will be fixed with #21461.

This issue won't be fixed by #21461.

> because we should not run any managed code after we called shutdown

We moved away from the concept of shutdown and no managed code running after shutdown. The runtime is not full cleaned up yet - there is still some of the shutdown code present. The shutdown should just send notifications that we are shutting down, but it should not otherwise prevent managed code from running.

My guess is that the GC hole you are seeing is caused by some part of the system "shutdown" and not working properly. You can try to understand the complex interaction that leads to this, but I do not think it would be a time well spent.

https://github.com/dotnet/coreclr/issues/17394#issuecomment-378517385 and #11404 talk about this same problem.

I think the best way to make a progress on this is to fix #17394 first.

**Comment by sandreenko** [comment]
There is no actual work in this issue for 3.0, x64 Checked Windows_NT GCStress=0xc do not fail anymore because `DisabledPollingIntervalMilliseconds` was set to 20 secs now.
#22159 tracks the issue that was found here.
